README.md
[comment]: # (This file is part of Gectrl, PHP Genereric controller. Copyright 2021 Kjell-Inge Gustafsson, kigkonsult, All rights reserved, licence LGPLv3)

# Gectrl

Gectrl is a PHP generic controller class package

* Supports the MVC software design pattern
* Distinguish controller and application logic using a strategy pattern
  
The controller
* provides coordination logic
* delegates application logic to actionClasses


* using implementations of the (strategy) [ActionClassInterface],
* invoking of actionClass condition **evaluate** and opt, logic **doAction** methods,
* passing all data information in an encapsulated [Package] class instance
  * input, output, config, logger etc


## Usage

#### Basic

Simpler (http/html) example

``` php
<?php
namespace Kigkonsult\Gectrl;
use ActionSrc\PrepAction;
use ActionSrc\CreateAction;
use ActionSrc\ReadAction;
use ActionSrc\UpdateAction;
use ActionSrc\DeleteAction;
use ActionSrc\CatchUpAction;
require 'vendor/autoload.php';

...

$package = Gectrl::init( $config, $logger )
    ->setActionClasses(
        [
            PrepAction::class,
            CreateAction::class,
            ReadAction::class,
            UpdateAction::class,
            DeleteAction::class,
            CatchUpAction::class,
        ]
    )
    ->main( $_REQUEST );
...

echo $package->getOutput();
```

For more detailed usage, read [Gectrl], [ActionClassInterface] and [Package] docs. 

## Installation

[Composer], from the Command Line:

```
composer require kigkonsult/gectrl
```

In your composer.json:

``` json
{
    "require": {
        "kigkonsult/gectrl": "dev-master"
    }
}
```

Version 1.8 supports PHP 8, 1.6 7.4, 1.4 7.0.

## Sponsorship
Donation using [paypal.me/kigkonsult] are appreciated.
For invoice, please email</a>.

## Licence

Gectrl is licensed under the LGPLv3 License.

[ActionClassInterface]:docs/ActionClassInterface.md
[Composer]:https://getcomposer.org/
[Gectrl]:docs/Gectrl.md
[paypal.me/kigkonsult]:https://paypal.me/kigkonsult
[Package]:docs/Package.md


releaseNotes.txt

1.2    PHP 7 (PHPStan checks)

1.6    PHP 7.4

1.8    PHP 8.0

1.8.1  Gectrl::main actionClasses sort moved to Gectrl::addActionClass
1.8.2  Package config/logger removed, used as new args
       at ActionClassInterface::evaluate/doAction invoke
1.8.3  Altered Package::getGuid generation
1.8.4  Gecrl Package generation moved to Gectrl::setInput
1.8.5  new Gecrl::processOne() and Gecrl::processOne() methods


test/autoload.php
<?php
/**
 * Gectrl, PHP generic controller
 *
 * This file is a part of Gectrl.
 *
 * @author    Kjell-Inge Gustafsson, kigkonsult <ical@kigkonsult.se>
 * @copyright 2021-22 Kjell-Inge Gustafsson, kigkonsult, All rights reserved
 * @link      https://kigkonsult.se
 * @license   Subject matter of licence is the software Gectrl.
 *            The above copyright, link, package and version notices,
 *            this licence notice shall be included in all copies or substantial
 *            portions of the Gectrl.
 *
 *            Gectrl is free software: you can redistribute it and/or modify
 *            it under the terms of the GNU Lesser General Public License as
 *            published by the Free Software Foundation, either version 3 of
 *            the License, or (at your option) any later version.
 *
 *            Gectrl is distributed in the hope that it will be useful,
 *            but WITHOUT ANY WARRANTY; without even the implied warranty of
 *            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *            GNU Lesser General Public License for more details.
 *
 *            You should have received a copy of the GNU Lesser General Public License
 *            along with Gectrl. If not, see <https://www.gnu.org/licenses/>.
 */
spl_autoload_register(
    function( string $class ) {
        static $BS      = '\\';
        static $PHP     = '.php';
        static $PREFIX  = 'Kigkonsult\\Gectrl\\';
        static $SRC     = 'src';
        static $SRCDIR  = null;
        static $TEST    = 'test';
        static $TESTDIR = null;
        if( is_null( $SRCDIR )) {
            $SRCDIR  = dirname( __DIR__ ) . DIRECTORY_SEPARATOR . $SRC . DIRECTORY_SEPARATOR;
            $TESTDIR = dirname(__DIR__ ) . DIRECTORY_SEPARATOR . $TEST . DIRECTORY_SEPARATOR;
        }
        if( 0 != strncmp( $PREFIX, $class, 18 ))
            return false;
        $class = substr( $class, 18 );
        if( false !== strpos( $class, $BS ))
            $class = str_replace( $BS, DIRECTORY_SEPARATOR, $class );
        $file = $SRCDIR . $class . $PHP;
        if( file_exists( $file )) {
            include $file;
        }
        else {
            $file = $TESTDIR . $class . $PHP;
            if( file_exists( $file )) {
                include $file;
            }
        }
    }
);


test/GectrlTest.php
<?php
/**
 * Gectrl, PHP generic controller
 *
 * This file is a part of Gectrl.
 *
 * @author    Kjell-Inge Gustafsson, kigkonsult <ical@kigkonsult.se>
 * @copyright 2021-22 Kjell-Inge Gustafsson, kigkonsult, All rights reserved
 * @link      https://kigkonsult.se
 * @license   Subject matter of licence is the software Gectrl.
 *            The above copyright, link, package and version notices,
 *            this licence notice shall be included in all copies or substantial
 *            portions of the Gectrl.
 *
 *            Gectrl is free software: you can redistribute it and/or modify
 *            it under the terms of the GNU Lesser General Public License as
 *            published by the Free Software Foundation, either version 3 of
 *            the License, or (at your option) any later version.
 *
 *            Gectrl is distributed in the hope that it will be useful,
 *            but WITHOUT ANY WARRANTY; without even the implied warranty of
 *            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *            GNU Lesser General Public License for more details.
 *
 *            You should have received a copy of the GNU Lesser General Public License
 *            along with Gectrl. If not, see <https://www.gnu.org/licenses/>.
 */
declare( strict_types = 1 );
namespace Kigkonsult\Gectrl;

use Exception;
use InvalidArgumentException;
use PHPUnit\Framework\TestCase;
use RuntimeException;

class GectrlTest extends TestCase
{
    private static string $VAR1 = 'var1';
    private static string $VAR2 = 'var2';
    private static string $VAR3 = 'var3';

    /**
     * Testing Gectrl and Package
     *
     * @test
     * @throws Exception
     */
    public function gectrlTest1() : void
    {
        $gectrl   = Gectrl::init( self::$VAR1, self::$VAR2 );
        $package1 = new Package();

        $this->assertTrue( $gectrl->isConfigSet(), 'error 41' );
        $this->assertEquals( self::$VAR1, $gectrl->getConfig(), 'error 142');

        $gectrl->setConfig( self::$VAR3 );
        $this->assertEquals( self::$VAR3, $gectrl->getConfig(), 'error 143');

        $this->assertTrue( $gectrl->isLoggerSet(), 'error 51' );
        $this->assertEquals( self::$VAR2, $gectrl->getLogger(), 'error 152');

        $gectrl->setLogger( self::$VAR3 );
        $this->assertEquals( self::$VAR3, $gectrl->getLogger(), 'error 153');

        $gectrl->setPackage( Package::init( self::$VAR3 ));
        $this->assertNotEquals(
            $package1->getTimestamp(),
            $gectrl->getPackage()->getTimestamp(),
            'error 161'
        );
        $this->assertNotEquals(
            $package1->getCorrelationId(),
            $gectrl->getPackage()->getCorrelationId(),
            'error 162'
        );

        $this->assertFalse(
            $gectrl->isActionClassSet(),
            'error 181'
        );

        $this->assertNotFalse(
            strpos(
                $package1->getLoadStatus(),
                $package1->getCorrelationId()
            ),
            'error 191'
        );
    }

    /**
     * Testing Gectrl, main, Package with no input
     *
     * @test
     * @throws Exception
     */
    public function gectrlTest2() : void
    {
        $detected = false;
        try {
            Gectrl::init()->setPackage( Package::init());
        }
        catch( InvalidArgumentException $ie ) {
            $detected = true;
        }
        $this->assertTrue( $detected, 'error 211' );
    }

    /**
     * Testing Gectrl::main
     *
     * @test
     * @throws Exception
     */
    public function gectrlTest3() : void
    {
        $actionClasses = [
            AcSrc\OtherInterface::class,
            AcSrc\OtherTrait::class,
            AcSrc\Action0Base::class,
            AcSrc\Action1Test::class,
            AcSrc\ActionExampleTest::class,
            AcSrc\Action3Test::class,
            AcSrc\Action4Test::class
        ];
        $gectrl   = Gectrl::init(
            null,
            null,
            $actionClasses
        );

        $this->assertTrue(
            $gectrl->isActionClassSet(),
            'error 311'
        );
        $this->assertTrue(
            $gectrl->isActionClassSet( AcSrc\ActionExampleTest::class ),
            'error 312a'
        );
        $this->assertFalse(
            $gectrl->isActionClassSet( AcSrc\OtherInterface::class ),
            'error 312b'
        );
        $this->assertFalse(
            $gectrl->isActionClassSet( AcSrc\OtherTrait::class ),
            'error 312c'
        );

        $this->assertNotCount(
            count( $actionClasses ), $gectrl->getActionClasses(), 'error 321'
        );

        $package = $gectrl->main( self::$VAR1 );

        $this->assertTrue( $package->isWorkDataKeySet( AcSrc\ActionExampleTest::class ), 'error 221' );
        $this->assertTrue( $package->isResultLogKeySet( AcSrc\ActionExampleTest::class ), 'error 222' );
        $this->assertEquals( $package->getInput(), $package->getOutput(), 'error 331' );
    }

    /**
     * Testing Gectrl:main assert 1, no input/no ActionClasses
     *
     * @test
     * @throws Exception
     * @throws Exception
     */
    public function gectrlTest4() : void
    {
        $detected = false;
        try {
            $package = Gectrl::init()->main();
        }
        catch( RuntimeException $re ) {
            $detected = true;
        }
        $this->assertTrue( $detected, 'error 411' );

        $detected = false;
        try {
            $package = Gectrl::init()->main( self::$VAR1 );
        }
        catch( RuntimeException $re ) {
            $detected = true;
        }
        $this->assertTrue( $detected, 'error 421' );
    }

    /**
     * Testing Gectrl, actionClass assert 2, invalid actionClasses
     *
     * @test
     * @throws Exception
     */
    public function gectrlTest5() : void
    {
        $detected = false;
        try {
            $gectrl = Gectrl::init(
                null,
                null,
                [ self::$VAR1 ]
            );
        }
        catch( InvalidArgumentException $ie ) {
                $detected = true;
        }
        $this->assertTrue( $detected, 'error 511' );

        $detected = false;
        try {
            $gectrl = Gectrl::init(
                null,
                null,
                [ AcSrc\OtherTest::class ]
            );
        }
        catch( InvalidArgumentException $ie ) {
            $detected = true;
        }
        $this->assertTrue( $detected, 'error 521' );
    }

    /**
     * Testing Gectrl, processOne (main)
     *
     * @test
     * @throws Exception
     */
    public function gectrlTest6() : void
    {
        $package1 = Package::init( self::$VAR1 );

        $package2 = Gectrl::init(
            null,
            null,
            [ AcSrc\ActionExampleTest::class ]
        )
            ->processOne( $package1 );

        $this->assertEquals( self::$VAR1, $package2->getOutput(), 'error 611');
    }

    /**
     * Testing Gectrl, processMany
     *
     * @test
     * @throws Exception
     */
    public function gectrlTest7() : void
    {
        $transactions = [
            self::$VAR1,
            self::$VAR2,
            self::$VAR3
        ];
        $input = [];
        foreach( $transactions as $trans ) {
            $input[] = Package::init( $trans );
        }

        $output = Gectrl::init(
            null,
            null,
            [ AcSrc\ActionExampleTest::class ]
        )
            ->processMany( $input );

        foreach( $transactions as $tix => $trans ) {
            $this->assertEquals( $trans, $output[$tix]->getOutput(), 'error 61' . $tix );
        }
    }
}


test/PackageTest.php
<?php
/**
 * Gectrl, PHP generic controller
 *
 * This file is a part of Gectrl.
 *
 * @author    Kjell-Inge Gustafsson, kigkonsult <ical@kigkonsult.se>
 * @copyright 2021-22 Kjell-Inge Gustafsson, kigkonsult, All rights reserved
 * @link      https://kigkonsult.se
 * @license   Subject matter of licence is the software Gectrl.
 *            The above copyright, link, package and version notices,
 *            this licence notice shall be included in all copies or substantial
 *            portions of the Gectrl.
 *
 *            Gectrl is free software: you can redistribute it and/or modify
 *            it under the terms of the GNU Lesser General Public License as
 *            published by the Free Software Foundation, either version 3 of
 *            the License, or (at your option) any later version.
 *
 *            Gectrl is distributed in the hope that it will be useful,
 *            but WITHOUT ANY WARRANTY; without even the implied warranty of
 *            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *            GNU Lesser General Public License for more details.
 *
 *            You should have received a copy of the GNU Lesser General Public License
 *            along with Gectrl. If not, see <https://www.gnu.org/licenses/>.
 */
declare( strict_types = 1 );
namespace Kigkonsult\Gectrl;

use Exception;
use Kigkonsult\KeyValueMgr\KeyValueMgr;
use PHPUnit\Framework\TestCase;

class PackageTest extends TestCase
{
    private static string $VAR1 = 'var1';
    private static string $VAR2 = 'var2';
    private static string $VAR3 = 'var3';

    /**
     * Testing Package
     *
     * @test
     * @throws Exception
     */
    public function packageTest1() : void
    {
        $package = Package::init( self::$VAR1 );

        $this->assertIsFloat( $package->getTimestamp(), 'error 21' );
        $timeStamp = microtime( true );
        $package->setTimestamp( $timeStamp );
        $this->assertIsFloat( $package->getTimestamp(), 'error 22' );
        $this->assertEquals(
            number_format( $timeStamp, 6 ),
            number_format( $package->getTimestamp(), 6 ),
            'error 23'
        );

        $this->assertIsString( $package->getCorrelationId(), 'error 31' );
        $package->setCorrelationId( self::$VAR3 );
        $this->assertEquals( self::$VAR3, $package->getCorrelationId(), 'error 32');

        $this->assertTrue( $package->isInputSet(), 'error 61' );
        $this->assertEquals( self::$VAR1, $package->getInput(), 'error 63');
        $package->setInput( self::$VAR3 );
        $this->assertEquals( self::$VAR3, $package->getInput(), 'error 63');

        $this->assertFalse( $package->isOutputSet(), 'error 71' );
        $package->setOutput( self::$VAR3 );
        $this->assertTrue( $package->isOutputSet(), 'error 72' );
        $this->assertEquals( self::$VAR3, $package->getOutput(), 'error 73');

        $this->assertInstanceOf(KeyValueMgr::class, $package->getWorkData(), 'error 80' );
        $this->assertFalse( $package->isWorkDataSet(), 'error 81' );
        $this->assertEquals( [], $package->getWorkDataKeys(), 'error 82' );
        $this->assertFalse( $package->getWorkData( self::$VAR3 ), 'error 83' );
        $this->assertFalse( $package->isWorkDataKeySet( self::$VAR3 ), 'error 84' );
        $package->setWorkData( self::$VAR3, self::$VAR2 );
        $this->assertTrue( $package->isWorkDataKeySet( self::$VAR3 ), 'error 85' );
        $this->assertEquals( self::$VAR2, $package->getWorkData( self::$VAR3 ), 'error 86' );

        $packageToString = $package->getLoadStatus();
        $this->assertStringContainsString( self::$VAR3, $packageToString, 'error 87' );
        $this->assertStringContainsString( self::$VAR2, $packageToString, 'error 88' );
        $package->getWorkData()->remove( self::$VAR3 );

        $this->assertInstanceOf(KeyValueMgr::class, $package->getResultLog(), 'error 90' );
        $this->assertFalse( $package->isResultLogSet(), 'error 91' );
        $this->assertEquals( [], $package->getResultLogKeys(), 'error 92' );
        $this->assertFalse( $package->getResultLog( self::$VAR3 ), 'error 93' );
        $this->assertFalse( $package->isResultLogKeySet( self::$VAR3 ), 'error 94' );
        $package->setResultLog( self::$VAR3, self::$VAR2 );
        $this->assertTrue( $package->isResultLogKeySet( self::$VAR3 ), 'error 95' );
        $this->assertEquals( self::$VAR2, $package->getResultLog( self::$VAR3 ), 'error 96' );

        $packageToString = $package->getLoadStatus();
        $this->assertStringContainsString( self::$VAR3, $packageToString, 'error 99' );
        $this->assertStringContainsString( self::$VAR2, $packageToString, 'error 98' );
    }
}


autoload.php
<?php
/**
 * Gectrl, PHP generic controller
 *
 * This file is a part of Gectrl.
 *
 * @author    Kjell-Inge Gustafsson, kigkonsult <ical@kigkonsult.se>
 * @copyright 2021-22 Kjell-Inge Gustafsson, kigkonsult, All rights reserved
 * @link      https://kigkonsult.se
 * @license   Subject matter of licence is the software Gectrl.
 *            The above copyright, link, package and version notices,
 *            this licence notice shall be included in all copies or substantial
 *            portions of the Gectrl.
 *
 *            Gectrl is free software: you can redistribute it and/or modify
 *            it under the terms of the GNU Lesser General Public License as
 *            published by the Free Software Foundation, either version 3 of
 *            the License, or (at your option) any later version.
 *
 *            Gectrl is distributed in the hope that it will be useful,
 *            but WITHOUT ANY WARRANTY; without even the implied warranty of
 *            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *            GNU Lesser General Public License for more details.
 *
 *            You should have received a copy of the GNU Lesser General Public License
 *            along with Gectrl. If not, see <https://www.gnu.org/licenses/>.
 */
spl_autoload_register(
    function( string $class ) {
        static $BS      = '\\';
        static $PHP     = '.php';
        static $PREFIX  = 'Kigkonsult\\Gectrl\\';
        static $SRC     = 'src';
        static $SRCDIR  = null;
        if( 0 != strncmp( $PREFIX, $class, 18 ))
            return false;
        $class = substr( $class, 18 );
        if( false !== strpos( $class, $BS ))
            $class = str_replace( $BS, DIRECTORY_SEPARATOR, $class );
        if( null === $SRCDIR ) {
            $SRCDIR  = __DIR__ . DIRECTORY_SEPARATOR . $SRC . DIRECTORY_SEPARATOR;
        }
        $file = $SRCDIR . $class . $PHP;
        if( file_exists( $file )) {
            include $file;
        }
    }
);


src/Util.php
<?php
/**
 * Gectrl, PHP generic controller
 *
 * This file is a part of Gectrl.
 *
 * @author    Kjell-Inge Gustafsson, kigkonsult <ical@kigkonsult.se>
 * @copyright 2021-22 Kjell-Inge Gustafsson, kigkonsult, All rights reserved
 * @link      https://kigkonsult.se
 * @license   Subject matter of licence is the software Gectrl.
 *            The above copyright, link, package and version notices,
 *            this licence notice shall be included in all copies or substantial
 *            portions of the Gectrl.
 *
 *            Gectrl is free software: you can redistribute it and/or modify
 *            it under the terms of the GNU Lesser General Public License as
 *            published by the Free Software Foundation, either version 3 of
 *            the License, or (at your option) any later version.
 *
 *            Gectrl is distributed in the hope that it will be useful,
 *            but WITHOUT ANY WARRANTY; without even the implied warranty of
 *            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *            GNU Lesser General Public License for more details.
 *
 *            You should have received a copy of the GNU Lesser General Public License
 *            along with Gectrl. If not, see <https://www.gnu.org/licenses/>.
 */
declare( strict_types = 1 );
namespace Kigkonsult\Gectrl;

use Exception;
use InvalidArgumentException;
use ReflectionClass;
use ReflectionException;

use function bin2hex;
use function chr;
use function ord;
use function random_bytes;
use function str_split;
use function strcasecmp;
use function vsprintf;

/**
 * Gectrl util, assert and sort methods
 */
class Util
{
    /**
     * @see https://www.php.net/manual/en/function.com-create-guid.php#117893
     * @return string
     * @throws Exception
     * @since 20220509 1.8.3
     */
    public static function getGuid() : string
    {
        static $FMT = '%s%s-%s-%s-%s-%s%s%s';
        $bytes = random_bytes( 16 );
        $bytes[6] = chr( ord( $bytes[6] ) & 0x0f | 0x40 ); // set version to 0100
        $bytes[8] = chr( ord( $bytes[8] ) & 0x3f | 0x80 ); // set bits 6-7 to 10
        return vsprintf( $FMT, str_split( bin2hex( $bytes ), 4 ) );
    }

    /**
     * Asserts actionClass
     *
     * Return true for class implementing ActionClassInterface,
     *   trait / interface / abstract class returns false
     * Throws exception on Reflection error
     *
     * @link https://www.php.net/manual/en/language.operators.type.php#102988
     * @param class-string<ActionClassInterface> $actionClass   ActionClassInterface class FQCN
     * @return bool
     * @throws InvalidArgumentException
     */
    public static function assertActionClass( string $actionClass ) : bool
    {
        static $FMT7 = 'Reflectionerror for %s : %s';
        static $FMT8 = 'Class %s do not implement ActionClassInterface';
        try {
            $reflectionClass = new ReflectionClass( $actionClass );
        }
        catch( ReflectionException $re ) {
            throw new InvalidArgumentException( sprintf( $FMT7, $actionClass, $re->getMessage() ), 21, $re );
        }
        return match ( true ) {
            ( $reflectionClass->isTrait() || // catch both false and null
                $reflectionClass->isInterface() || $reflectionClass->isAbstract() ) => false,
            $reflectionClass->implementsInterface( ActionClassInterface::class ) => true,
            default => throw new InvalidArgumentException( sprintf( $FMT8, $actionClass ), 28 )
        };
    }
    /**
     * @var callable
     */
    public static $ACTIONCLASSSORTER = [ __CLASS__, 'actionClassSort' ];

    /**
     * Sort ActionClassInterfaces (i.e. string FQCNs) on method getExecOrder result
     *
     * If equal, sort on actionClasses name
     *
     * @param class-string<ActionClassInterface> $a
     * @param class-string<ActionClassInterface> $b
     * @return int
     */
    public static function actionClassSort( string $a, string $b ) : int
    {
        $aeo = $a::getExecOrder();
        $beo = $b::getExecOrder();
        return match( true ) {
            ( $aeo === $beo ) => strcasecmp( $a, $b ),
            ( $aeo < $beo )   => -1,
            default           => 1
        };
    }
}


src/Gectrl.php
<?php
/**
 * Gectrl, PHP generic controller
 *
 * This file is a part of Gectrl.
 *
 * @author    Kjell-Inge Gustafsson, kigkonsult <ical@kigkonsult.se>
 * @copyright 2021-22 Kjell-Inge Gustafsson, kigkonsult, All rights reserved
 * @link      https://kigkonsult.se
 * @license   Subject matter of licence is the software Gectrl.
 *            The above copyright, link, package and version notices,
 *            this licence notice shall be included in all copies or substantial
 *            portions of the Gectrl.
 *
 *            Gectrl is free software: you can redistribute it and/or modify
 *            it under the terms of the GNU Lesser General Public License as
 *            published by the Free Software Foundation, either version 3 of
 *            the License, or (at your option) any later version.
 *
 *            Gectrl is distributed in the hope that it will be useful,
 *            but WITHOUT ANY WARRANTY; without even the implied warranty of
 *            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *            GNU Lesser General Public License for more details.
 *
 *            You should have received a copy of the GNU Lesser General Public License
 *            along with Gectrl. If not, see <https://www.gnu.org/licenses/>.
 */
declare( strict_types = 1 );
namespace Kigkonsult\Gectrl;

use Exception;
use InvalidArgumentException;
use RuntimeException;

use function in_array;
use function sprintf;
use function usort;

/**
 * Class Gectrl
 *
 * Gectrl is a PHP generic controller class
 *   Supports the MVC software design pattern
 *   Distinguish controller and application logic using a strategy pattern
 *
 * The controller provides coordination logic
 *
 * The controller delegates application logic to actionClasses,
 *   using implementations of the (strategy) ActionClassInterface,
 *   invoking of actionClass condition 'evaluate' and logic 'doAction' methods,
 *   passing all data information in an encapsulated Package class instance
 *     along with opt config and logger
 *
 * @package Kigkonsult\Gectrl
 * @since 20220511 1.8.5
 */
class Gectrl
{
    /**
     * Array (string[]) FQCNs for actionClasses
     *
     * @var class-string<ActionClassInterface>[]
     */
    private array $actionClasses = [];

    /**
     * @var bool
     */
    private bool $actionClassesCheck = false;

    /**
     * @var null|Package
     */
    private ? Package $package = null;

    /**
     * Opt any config
     *
     * @var null|mixed
     * @since 20220509 1.8.2
     */
    private mixed $config = null;

    /**
     * Opt any logger
     *
     * @var null|mixed
     * @since 20220509 1.8.2
     */
    private mixed $logger = null;

    /**
     * Gectrl constructor
     *
     * @param mixed $config
     * @param mixed $logger
     * @param class-string<ActionClassInterface>[] $actionClasses
     * @throws Exception
     * @since 20220510 1.8.4
     */
    public function __construct(
        mixed $config = null,
        mixed $logger = null,
        ? array $actionClasses = []
    )
    {
        if( null !== $config ) {
            $this->config = $config;
        }
        if( null !== $logger ) {
            $this->logger = $logger;
        }
        if( ! empty( $actionClasses )) {
            $this->setActionClasses( $actionClasses );
        }
    }

    /**
     * Gectrl factory method
     *
     * @param mixed    $config
     * @param mixed    $logger
     * @param class-string<ActionClassInterface>[] $actionClasses
     * @return Gectrl
     * @throws Exception
     * @since 20220509 1.8.2
     */
    public static function init(
        mixed $config = null,
        mixed $logger = null,
        ? array $actionClasses = []
    ) : Gectrl
    {
        return new self( $config, $logger, $actionClasses );
    }

    /**
     * Process ONE input transaction
     *
     * @param mixed $input   any kind of input (scalar/array/object) or a Package instance
     * @return Package
     * @throws Exception
     * @throws RuntimeException
     * @since 20220511 1.8.5
     */
    public function processOne( mixed $input ) : Package
    {
        $this->setInput( $input );
        return $this->main();
    }

    /**
     * Process MULTIPLE input transactions
     *
     * Empty input array allowed, returns empty array
     *
     * @param mixed[]|Package[] $input   any kind of input, array of scalar/array/object or Package[]
     * @return Package[]
     * @throws Exception
     * @throws RuntimeException
     * @since 20220511 1.8.5
     */
    public function processMany( array $input ) : array
    {
        $output = [];
        foreach( $input as $packageInput ) {
            $output[] = $this->processOne( $packageInput );
        }
        return $output;
    }

    /**
     * Main method, assert Gectrl instance, invoke actionClass:evaluate/doAction in order
     *
     * Accepts any kind of input (scalar/array/object)
     *   for the default (internally) created Package class instance
     * OR
     *   a (replacing) externally created package class instance
     *
     * @param mixed $input   any kind of input, scalar/array/object or a Package instance
     * @return Package
     * @throws InvalidArgumentException
     * @throws RuntimeException
     * @since 20220511 1.8.5
     */
    public function main( mixed $input = null ) : Package
    {
        static $FMT1 = 'No Gectrl (Package input) set';
        static $FMT2 = 'No Gectrl actions set';
        if( null !== $input ) {
            $this->setInput( $input );
        }
        elseif(( null === $this->package )) {
            throw new RuntimeException( $FMT1, 11 );
        }
        if( ! $this->actionClassesCheck && ! $this->isActionClassSet()) {
            throw new RuntimeException( $FMT2, 12 );
        }
        foreach( $this->actionClasses as $actionClass ) {
            if( false === $actionClass::{ActionClassInterface::EVALUATE}(
                $this->package,
                $this->config,
                $this->logger
                )) {
                continue; // skip this action
            }
            if( true === $actionClass::{ActionClassInterface::DOACTION}(
                $this->package,
                $this->config,
                $this->logger
                )) {
                break; // true force break exec of actions
            }
        } // end foreach
        return $this->package;
    }

    /**
     * Return (string[]) actionClasses (FQCNs) sorted om execOrder (and FQCN)
     *
     * @return class-string<ActionClassInterface>[]  i.e. ActionClassInterface class FQCN list
     */
    public function getActionClasses() : array
    {
        return $this->actionClasses;
    }

    /**
     * Return bool, true if actionsClasses (or fqcn) is set, otherwise false
     *
     * @param string|null $fqcn
     * @return bool
     */
    public function isActionClassSet( ? string $fqcn = null ) : bool
    {
        if( ! empty( $fqcn )) {
            return in_array( $fqcn, $this->actionClasses, true );
        }
        if( ! empty( $this->actionClasses )) {
            $this->actionClassesCheck = true;
            return true;
        }
        return false;
    }

    /**
     * Add (string) actionClass (FQCN)
     *
     * Opt. traits / interfaces / abstract classes are ignored
     * Throws exception on Reflection error
     *
     * @param class-string<ActionClassInterface> $actionClass   ActionClassInterface class FQCN
     * @return Gectrl
     * @throws InvalidArgumentException
     * @since 20220509 1.8.1
     */
    public function addActionClass( string $actionClass ) : Gectrl
    {
        if( Util::assertActionClass( $actionClass ) ) {
            $this->actionClasses[] = $actionClass;
            usort( $this->actionClasses, Util::$ACTIONCLASSSORTER );
        }
        return $this;
    }

    /**
     * Set (string[]) actionClasses (FQCNs)
     *
     * For extracting namespace(s) for actionsClasses,
     * you may use https://gitlab.com/hpierce1102/ClassFinder
     *
     * @param class-string<ActionClassInterface>[] $actionClasses  ActionClassInterface class FQCN list
     * @return Gectrl
     * @throws InvalidArgumentException
     */
    public function setActionClasses( array $actionClasses ) : Gectrl
    {
        $this->actionClasses = [];
        foreach( $actionClasses as $actionFqcn ) {
            $this->addActionClass( $actionFqcn );
        }
        return $this;
    }

    /**
     * Return Package
     *
     * @return null|Package
     */
    public function getPackage() : ? Package
    {
        return $this->package;
    }

    /**
     * Set (replace) Package
     *
     * @param mixed|Package $input
     * @throws InvalidArgumentException
     * @return Gectrl
     * @since 20220511 1.8.5
     */
    public function setInput( mixed $input ) : Gectrl
    {
        if( ! $input instanceof Package ) {
            try {
                $input = new Package( $input );
            }
            catch( Exception $e ) {
                throw new InvalidArgumentException( $e->getMessage(), 31, $e );
            }
        }
        $this->setPackage( $input );
        return $this;
    }

    /**
     * Set (replace) Package, input MUST be set
     *
     * @param Package $package
     * @throws InvalidArgumentException
     * @return Gectrl
     * @since 20220511 1.8.5
     */
    public function setPackage( Package $package ) : Gectrl
    {
        static $FMT4 = 'No input set in Package %s / %f';
        if( ! $package->isInputSet()) {
            throw new InvalidArgumentException(
                sprintf( $FMT4, $package->getCorrelationId(), $package->getTimestamp()),
                41
            );
        }
        $this->package = $package;
        return $this;
    }

    /**
     * Return (mixed) config
     *
     * @return mixed
     */
    public function getConfig() : mixed
    {
        return $this->config;
    }

    /**
     * Return bool, true if config is set, otherwise false
     *
     * @return bool
     */
    public function isConfigSet() : bool
    {
        return ( null !== $this->config );
    }

    /**
     * Set (mixed) config
     *
     * @param mixed $config
     * @return Gectrl
     */
    public function setConfig( mixed $config ) : Gectrl
    {
        $this->config = $config;
        return $this;
    }

    /**
     * Return logger
     *
     * @return mixed
     */
    public function getLogger() : mixed
    {
        return $this->logger;
    }

    /**
     * Return bool, true if logger is set, otherwise false
     *
     * @return bool
     */
    public function isLoggerSet() : bool
    {
        return ( null !== $this->logger );
    }

    /**
     * Set logger
     *
     * @param mixed $logger
     * @return Gectrl
     */
    public function setLogger( mixed $logger ) : Gectrl
    {
        $this->logger = $logger;
        return $this;
    }
}
